numpy/__init__.pyi

```diff
@@ -386,133 +388,172 @@ class generic(_ArrayOrScalarCommon):
+    def real(self: _ArraySelf) -> _ArraySelf: ...
+    @property
+    def imag(self: _ArraySelf) -> _ArraySelf: ...
 
-class datetime64:
+class datetime64(generic):
     @overload
     def __init__(
         self,
-        __value: Union[None, datetime64, str, dt.datetime] = ...,
-        __format: str = ...
+        __value: Union[None, datetime64, str, bytes, dt.datetime] = ...,
+        __format: Union[str, bytes] = ...,
     ) -> None: ...
     @overload
-    def __init__(self, __value: int, __format: str) -> None: ...
+    def __init__(self, __value: int, __format: Union[str, bytes]) -> None: ...
     def __add__(self, other: Union[timedelta64, int]) -> datetime64: ...
     def __sub__(self, other: Union[timedelta64, datetime64, int]) -> timedelta64: ...
+    def __rsub__(self, other: Union[datetime64, int]) -> timedelta64: ...
 
-class integer(number, _real_generic): ...  # type: ignore
+class integer(number): ...  # type: ignore
 class signedinteger(integer): ...  # type: ignore
 
 class int8(signedinteger):
-    def __init__(self, __value: SupportsInt = ...) -> None: ...
+    def __init__(
+        self, __value: Union[SupportsInt, SupportsIndex, str, bytes] = ...
+    ) -> None: ...
 
 class int16(signedinteger):
-    def __init__(self, __value: SupportsInt = ...) -> None: ...
+    def __init__(
+        self, __value: Union[SupportsInt, SupportsIndex, str, bytes] = ...
+    ) -> None: ...
 
 class int32(signedinteger):
-    def __init__(self, __value: SupportsInt = ...) -> None: ...
+    def __init__(
+        self, __value: Union[SupportsInt, SupportsIndex, str, bytes] = ...
+    ) -> None: ...
 
 class int64(signedinteger):
-    def __init__(self, __value: SupportsInt = ...) -> None: ...
+    def __init__(
+        self, __value: Union[SupportsInt, SupportsIndex, str, bytes] = ...
+    ) -> None: ...
 
 class timedelta64(signedinteger):
-    def __init__(self, __value: Any = ..., __format: str = ...) -> None: ...
+    def __init__(
+        self,
+        __value: Union[None, int, str, bytes, dt.timedelta, timedelta64] = ...,
+        __format: Union[str, bytes] = ...,
+    ) -> None: ...
     @overload
     def __add__(self, other: Union[timedelta64, int]) -> timedelta64: ...
     @overload
     def __add__(self, other: datetime64) -> datetime64: ...
     def __sub__(self, other: Union[timedelta64, int]) -> timedelta64: ...
-    if sys.version_info[0] < 3:
-        @overload
-        def __div__(self, other: timedelta64) -> float: ...
-        @overload
-        def __div__(self, other: float) -> timedelta64: ...
+    def __rsub__(self, other: Union[timedelta64, int]) -> timedelta64: ...
     @overload
     def __truediv__(self, other: timedelta64) -> float: ...
     @overload
     def __truediv__(self, other: float) -> timedelta64: ...
+    def __rtruediv__(self, other: timedelta64) -> float: ...
     def __mod__(self, other: timedelta64) -> timedelta64: ...
+    def __rmod__(self, other: timedelta64) -> timedelta64: ...
```

Comment:
These `rop` overloads look useless to me - the cases where they are relevant already match the `__op__` overloads.

Reply:
Removing them seems to have no effect on the tests either (see https://github.com/numpy/numpy/pull/17214/commits/1276b4ba608065cb030cf601eb719f7b4ca45ba8).
The only exception here was `datetime64.__rsub__`, which triggered a fallback to `_ArrayOrScalarCommon.__sub__` or `.__rsub__` when removed (which is bad). For now I've left it in.